packages/core-ethereum/crates/core-ethereum-rpc/src/lib.rs

```diff
@@ -63,45 +87,96 @@ impl From<ethers::types::Log> for Log {
+#[derive(Debug, Clone)]
+pub struct BlockWithLogs {
+    /// Block with TX hashes.
+    pub block: Block,
+    /// Logs of interest corresponding to the block.
+    pub logs: Vec<Log>,
+}
+
+impl Display for BlockWithLogs {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{} and {} logs", self.block, self.logs.len())
+    }
+}
+
 /// Represents a query to extract logs containing specific contract events.
 #[derive(Debug, Clone)]
 pub struct EventsQuery {
     /// Contract address
     pub address: Address,
-    /// Event topics
+    /// Event topics (maximum 4)
     pub topics: Vec<TxHash>,
-    /// Start block number
-    pub from: u64,
-    /// End block number
-    pub to: u64,
 }
 
-#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
-#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
-pub trait HoprRpcOperations {
-    type BlockStream: Stream<Item = Block>;
+impl From<EventsQuery> for ethers::types::Filter {
+    fn from(value: EventsQuery) -> Self {
+        let addr: ethers::types::H160 = value.address.into();
+        let mut ret = ethers::types::Filter::new().address::<ethers::types::H160>(addr.into());
 
-    type LogStream: Stream<Item = Log>;
-
-    async fn genesis_block(&self) -> Result<u64>;
+        for i in 0..4.min(value.topics.len()) {
+            ret.topics[i] = Some(value.topics[i].into())
+        }
 
-    async fn block_number(&self) -> Result<u64>;
+        ret
+    }
+}
 
+/// Trait defining general set of operations an RPC provider
+/// must provide to the HOPR node.
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+pub trait HoprRpcOperations {
+    /// Retrieves the timestamp from the given block number.
     async fn get_timestamp(&self, block_number: u64) -> Result<Option<u64>>;
 
+    /// Retrieves the node's account balance of the given type.
     async fn get_balance(&self, balance_type: BalanceType) -> Result<Balance>;
 
-    async fn get_transactions_in_block(&self, block_number: u64) -> Result<Vec<Hash>>;
-
+    /// Retrieves info of the given node module's target.
     async fn get_node_management_module_target_info(&self, target: Address) -> Result<Option<U256>>;
 
+    /// Retrieves safe address of the given node address from the registry.
     async fn get_safe_from_node_safe_registry(&self, node: Address) -> Result<Address>;
 
+    /// Retrieves target address of the node module.
     async fn get_module_target_address(&self) -> Result<Address>;
 
+    /// Sends transaction to the RPC provider.
     async fn send_transaction(&self, tx: TypedTransaction) -> Result<Hash>;
+}
 
-    async fn subscribe_blocks(&self) -> Result<Self::BlockStream>;
+/// Extension of `HoprRpcOperations` trait with functionality required by the Indexer.
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+pub trait HoprIndexerRpcOperations: HoprRpcOperations {
```

Comment:
Should this not be reversed? I.e. have the trait be standalone... and then collect all api traits inside the `HoprRpcOperations : HoprIndexRpcOps, HoprTransactionRpcOps,...`